qutebrowser/completion/models/listcategory.py

```diff
@@ -61,9 +61,7 @@ def set_pattern(self, val):
             log.completion.warning(f"Trimming {len(val)}-char pattern to 5000")
             val = val[:5000]
         self._pattern = val
-        val = re.sub(r' +', r' ', val)  # See #1919
-        val = re.escape(val)
-        val = val.replace(r'\ ', '.*')
+        val = '(?=.*' + ')(?=.*'.join(map(re.escape, val.split())) + ')'
```

Comment:
I hope you understand how this works, because I have a couple of questions.

I understand that lookaheads usually are used relative to other patterns outside the lookahead, but since there is nothing outside the lookaheads they all run across the whole string independently. At least I think that's what enables the order independent matching.

So given I'm following along so far, my questions are:

1. why is the `.*` necessary in the lookaheads? I guess that they are since I tried removing them and it doesn't work. But we are only interested in looking for the search terms themselves, not a preceding any amount of any character. Why and how do they make it work?
2. my other questions were about in the completion widget only the search terms are highlighted, and not any preceding chars, but it looks like the highlighter is a separate regex. A much simple one like `foo|bar`, which does a very similar thing to this one but in that case only one of the search terms need to match, here we want them all to.

I believe you can make the `.*` bits non-greedy (`.*?`) to make them stop matching earlier, regex101 says it reduces the number of steps a bunch. And your tests still pass.

Reply:
I also wouldn't mind some documentation here to help people who aren't familiar with regex maintain this piece of code. I've certainly be staring at it for too long already. Maybe something like:

```python
# Positive lookahead per search term. This means that all search terms must be matched but they can be matched anywhere in the string, so they can be in any order. 
# For example "foo bar" -> `(?=.*?foo)(?=.*?bar)`
```

Reply:
> I hope you understand how this works, because I have a couple of questions.

Not exactly, but I'll try.

> 1. why is the `.*` necessary in the lookaheads? I guess that they are since I tried removing them and it doesn't work. But we are only interested in looking for the search terms themselves, not a preceding any amount of any character. Why and how do they make it work?

Otherwise the next searched word would have to be immediately after the first searched word?

> I believe you can make the `.*` bits non-greedy (`.*?`) to make them stop matching earlier, regex101 says it reduces the number of steps a bunch. And your tests still pass.

I would guess it doesn't matter when it's not in a capturing group. I didn't detect differences in the benchmarks.

@go2bodhi: comments?

Reply:
>> I believe you can make the .* bits non-greedy (.*?) to make them stop matching earlier, regex101 says it reduces the number of steps a bunch. And your tests still pass.

> I would guess it doesn't matter when it's not in a capturing group. I didn't detect differences in the benchmarks.

Hmm, no change for the better in the benchmarks for me either. Even with really long URLs, which is where regex101 says it does worse. Looks like it makes them worse if anything. Oh well.

>> why is the .* necessary in the lookaheads?

> Otherwise the next searched word would have to be immediately after the first searched word?

Okay, I think I'm kinda getting it now. I was thinking that since the groups can be matched in any order then they don't depend on the preceding group. But I guess once the groups are matched then the whole regex needs to apply against the same part of a string in some way. So it's less "find these terms in any order" and more "find a piece of text that contains these terms in any order".
[This answer](https://stackoverflow.com/a/470602) explains this as "the match position is reset", but I'm not sure that fully covers the implications of adjacent lookaheads.